### DISCUSSION TITLE
Validate my understanding of client object relationship to connection pools?

### DISCUSSION THREAD

**Discussion by rickbatka** [prisma/prisma#17788] [Q&A] [answered]
Hi, really loving Prisma so far, thanks for such a great tool! I just want to verify my understanding before I jump into the deep end with some advanced usages. Thanks in advance for any help or clarification you can provide.

The ["Connection Management" docs](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#long-running-processes) state that you should strive to only have one Prisma client object in your application, and that behind the scenes Prisma will manage a connection pool for that object. 

That makes sense - however...

I have a multitenant web application with a small number of tenants (less than 10). I use Postgres's row-level-security (RLS) feature to ensure data doesn't leak between my tenants. 

I wish to use a separate connection pool per tenant in my web application. I am aware of [this example](https://github.com/prisma/prisma-client-extensions/tree/main/examples/row-level-security), which wraps every query in a transaction in order to set a SQL variable that lasts the length of the transaction, but that approach is not appropriate for my use case for several reasons.

My plan is simple, and I just want someone to verify my understanding of how Prisma manages connection pools and confirm this will work:

My plan is to create one Prisma client object per tenant, and one for "empty" tenant ID. In my web application's request handler, I'll use the connection with the "empty" tenant ID to verify the user session and get the tenant ID the user is associated with. 

Then, every subsequent place in my application that needs to run a query, will use the Prisma client object I created (and cached) for that tenant. I don't mind manually managing which connection to use - for example, my user session verification function will call something like `getEmptyTenantPrismaClient().find(...)`, and later in my program, subsequent queries will call something like `getPrismaClientForTenant(tenantId).find(...)`.

I can happily manage the singleton logic to have my code always get the proper tenant's client object when I need a Prisma client.

My questions are:
1. It is my understanding that a Prisma client object maps 1-to-1 to a connection pool. **Is that accurate?** If I create several Prisma client objects, can I reliably manage a "pool of pools" as I described above?
2. Is there a way to set a connection-scoped variable for each of my pools? I thought I could just eagerly call `connect()` on the Prisma client object and then run Postgres's built-in `set_config()` function, but that function sets variables for a **session** (connection) - I worry that subsequent queries using that same Prisma client object might create new connections for the pool and skip that bit of setup. Is there no way to set a bit of state on a Prisma client object that affects every connection in the pool?
3. Given how I think the answer to my previous question is "no, you can't share state across all connections in a pool", I assume I'll need to create a DB user per tenant. This is fine, but I just want to be sure it's the safest way.

**Comment by janpio** [comment, marked as answer]
Re 1: Yes.
Re 2: No, you use Client Extensions for that. That is the only way right now to ensure that certain queries have certain state available when run via a specific Prisma Client.The RLS example you linked to does exactly that.
Re 3: When you have 1 Prisma Client per tenant, there is no chance that something leaks across those. You can only accidentally run queries without setting the correct session variables first - which is what Prisma Client Extensions can help avoid.

The big problem with the "1 Prisma Client per tenant" approach are your open database connections. If each Prisma Client can open 10 connections, that means e.g. 10x10 connections open all the time. If each deployment (and preview deployment) again does that, you can easily use up all open connections of your database. They are usually a surprisingly scarce resource.

As using separate connection pool (via separate Prisma Clients) does not really give you much more functionality with Prisma (you still need to use a Client Extension to make sure the correct session vars are set - so you can do the same with just one connection pool), we usually recommend to avoid that.

**Reply by rickbatka** [reply]
Thanks so much. I understand DB connections are very scarce indeed - in Postgres, I believe it's usually 2x the CPU count on the machine.

Still, I'm hesitant to follow the linked RLS example because I fear wrapping every query in a transaction might hurt performance. I'll have to try it on a synthetic dataset with some real-world access patterns and measure it. Also, we use batch transactions in a few places, and I don't think the query extensions preview yet has the necessary plumbing to make that work.

I also think that going with the pool-of-pools approach and a separate DB user for each pool, I will have an easier time if I decide to move to pgbouncer in the future, since it by default creates a new pool for every (user, database) tuple.

Thanks for the advice! I'll watch my connection counts.

**Reply by janpio** [reply]
As Prisma has no way to configure the session settings of a connection in Prisma Client, and opens connections when and how it deems necessary, I don't think you have another way than to set the necessary session settings for each query, and a transaction is the only safe way to do that. Happy to be corrected here of course if I am overlooking something.

**Reply by rickbatka** [reply]
Oh, I think the only way my solution works without session-scoped settings is to use different users (and therefore, different `datasource`s in the different prisma client objects), and enforce RLS based on the current DB user.

I guess I should explicitly state that I assumed the `datasource` URL that I pass to new Prisma client objects is what you're keying off to decide how to group pooled connections. If I create two prisma client objects, each with different DB usernames in their respective connection strings, this should work, right? Those pools will be forever separated?

**Reply by janpio** [reply]
Ah, I missed the "and enforce RLS based on the current DB user" bit before. With that it all makes sense.

> If I create two prisma client objects, each with different DB usernames in their respective connection strings, this should work, right? Those pools will be forever separated?

Yes, these will be two separate instances of Prisma Client with their own Query Engine which opens and manages the connections. No cross communication happening here at all.

**Reply by rickbatka** [reply]
Great, glad my reading of the docs was correct! Thanks for helping me get to the bottom of it.

Some way to achieve session-scoped SQL variables at the Prisma client object level would be a pretty sweet feature for this use case (and I'm sure there are others).  That said, the user-per-tenant workaround seems pretty clean for my use-case of having very few, seldom added tenants.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
